decisions/0003-files-in-JSON-output.md

```diff
@@ -121,27 +144,183 @@ lost
-However this will give Form Builder full control of access to files stored.
+- If we encrypt the files in S3 and also serve them encrypted, then we don't
+need to worry as much about bucket-level access to the files.
+
+- A proxy application is needed if the adapter does not have the responsibility
+of decrypting the files. A proxy would not be needed if the responsibility of
+decryption is moved to the adapter. However some kind of pre-shared key
+mechanism would be needed.
+
+- This also follows the principle of all user-submitted data in Form Builder
+always being encrypted for the next thing that needs it - user, then submitter,
+then receiver
+
+- It would probably be fine to use native S3 signed URLs that are valid for 1
+week, because if the URL is exposed then the file you can fetch from it is
+encrypted so we can worry less about expiring the URLs as soon as possible
+
+- Alerting on unusual patterns of requests on signed URLs might still be nice to
+have, but that's getting into anomaly detection which isn't that easy to do and
+may be of little benefit if we go down this route compared to others
+
+- How long we need or want to keep files retrievable for depends on whether
+they're served encrypted or not, how long retrieval will take after submission,
+and whether it's feasible to expire links after a single use or when Form
+Builder is notified that the file has been successfully retrieved and stored
+elsewhere as knowing we've served the file successfully doesn't guarantee the
+other side has it.
+
+### Issues concerning the revoking if signed URLs:
+
+S3 signed URLs expire after a given time, and not after a number of uses - if we
+want to do the latter we would have to implement a proxy application
+
+Revoking a S3 signed URL on demand involves removing the permissions of the IAM
+user who signed the URL, so we couldn't expire individual URLs ourselves after
+we know they've been used successfully unless we also generate a new IAM user to
+sign each URL for each file. This is presumably possible but not ideal. So
+revoking signed URLs on demand may also mean using a proxy application.
+
+CMSs which retrieve files directly themselves are unlikely to be able to notify
+us when they've successfully fetched the file - a decrypting adapter could make
+this callback, but probably wouldn't know whether the CMS had stored the file.
+
+This whole question means that Form Builder needs not to care too much about
+what's happening a long way downstream.
+
+### Conclusion
+
+Considering all the points above option 6 has been chosen.
+
+It leaves a clear cut boundary between Form Builder and integrations. Although
+the responsibility of decrypting files is now in the hands of the receiver. The
+Form Builder team does not need a proxy application and S3 can be left to handle
+files.
+
+It deals with encryption: files are kept encrypted at rest and only decrypted
+when being processed or exiting the system.
```

Comment:
We'd be serving files still encrypted so they'd only be decrypted during submission processing inside FB, and by the receiver outside the control of the platform.

Reply:
Yes, I have updated accordingly. This assumes the adapter is not part of the Form Builder ecosystem